prescient/plugins/tracking.py

```diff
@@ -0,0 +1,1619 @@
+            else:
+                return m.start_up[j,h] - m.shut_dw[j,h] == m.on_off[j,h] - m.on_off[j,h-1]
+        m.start_up_shut_dw = pyo.Constraint(m.UNITS,m.HOUR,rule = start_up_shut_dw_fun)
+
+        # either start up or shut down
+        def start_up_or_shut_dw_fun(m,j,h):
+            return m.start_up[j,h] + m.shut_dw[j,h] <= 1
+        m.start_up_or_shut_dw = pyo.Constraint(m.UNITS,m.HOUR,rule = start_up_or_shut_dw_fun)
+
+        # ramp up limits
+        def ramp_up_fun(m,j,h):
+            '''
+            j,h stand for unit, hour,scenario respectively.
+            '''
+            if h==0:
+                return m.P_T[j,h] <= m.pre_P_T[j] \
+                + m.ramp_up[j]*m.pre_on_off[j]\
+                + m.ramp_start_up[j]*m.start_up[j,h]
+            else:
+                return m.P_T[j,h] <= m.P_T[j,h-1] \
+                + m.ramp_up[j]*m.on_off[j,h-1]\
+                + m.ramp_start_up[j]*m.start_up[j,h]
+        m.ramp_up_con = pyo.Constraint(m.UNITS,m.HOUR,rule = ramp_up_fun)
+
+        # ramp shut down limits
+        def ramp_shut_dw_fun(m,j,h):
+            '''
+            j,h stand for unit, hour,scenario respectively.
+            '''
+            if h==0:
+                return m.pre_P_T[j] <= m.Pmax[j]*m.on_off[j,h] + m.ramp_shut_dw[j] * m.shut_dw[j,h]
+            else:
+                return m.P_T[j,h-1] <= m.Pmax[j]*m.on_off[j,h] + m.ramp_shut_dw[j] * m.shut_dw[j,h]
+        m.ramp_shut_dw_con = pyo.Constraint(m.UNITS,m.HOUR,rule = ramp_shut_dw_fun)
+
+        # ramp down limits
+        def ramp_dw_fun(m,j,h):
+            '''
+            j,h stand for unit, hour,scenario respectively.
+            '''
+            if h == 0:
+                return m.pre_P_T[j] - m.P_T[j,h] <= m.ramp_dw[j] * m.on_off[j,h]\
+                + m.ramp_shut_dw[j] * m.shut_dw[j,h]
+            else:
+                return m.P_T[j,h-1] - m.P_T[j,h] <= m.ramp_dw[j] * m.on_off[j,h]\
+                + m.ramp_shut_dw[j] * m.shut_dw[j,h]
+        m.ramp_dw_con = pyo.Constraint(m.UNITS,m.HOUR,rule = ramp_dw_fun)
+
+        ## add min up and down time constraints
+        self._add_UT_DT_constraints(m)
+
+        ## Expression
+
+        def prod_cost_fun(m,j,h):
+            return m.min_load_cost[j] * m.on_off[j,h] \
+            + sum(m.F[j,l]*m.power_segment[j,h,l] for l in m.SEGMENTS)
+        m.prod_cost_approx = pyo.Expression(m.UNITS,m.HOUR,rule = prod_cost_fun)
+
+        # start up costs
+        def start_cost_fun(m,j,h):
+            return m.start_up_cost[j]*m.start_up[j,h]
+        m.start_up_cost_expr = pyo.Expression(m.UNITS,m.HOUR,rule = start_cost_fun)
+
+        # total cost
+        def tot_cost_fun(m,j,h):
+            return m.prod_cost_approx[j,h] + m.start_up_cost_expr[j,h]
+        m.tot_cost = pyo.Expression(m.UNITS,m.HOUR,rule = tot_cost_fun)
+
+        return m
+
+    @staticmethod
+    def _add_UT_DT_constraints(m):
```

Comment:
Is this the same code as the bidding file? If yes, does it make sense to use inheritance? You could have both bidding and tracking inherit from a `Model_Builder` class. Only do this if the abstraction is easy.